### PR TITLE
Add cupy backend support for `scipy.ndimage`

### PR DESCRIPTION
**Pull request by Smit-create**
`uarray` support for `scipy.ndimage` is implemented in: https://github.com/scipy/scipy/pull/14356

This PR aims to add `cupy` backend for `scipy.ndimage`.

cc @rgommers

### CODE REVIEW

**Comment by Smit-create** [comment]
@rgommers @grlee77 Can you please have a look at the initial implementation? If this looks good then, I can start adding more functions to the backend.

**Comment by rgommers** [comment]
That looks like a good start to me. Matches the `cupyx.scipy.fft` implementation pattern, which I'd expect to be fine. Note that there are tests for the `fft` backend in https://github.com/cupy/cupy/blob/master/tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py#L151, adding similar ones for `ndimage` would be good.

**Comment by Smit-create** [comment]
> adding similar ones for `ndimage` would be good.

That sounds good but I think it would be more effective once we have https://github.com/scipy/scipy/pull/14356 merged (because we need `set_backend` from `scipy.ndimage`)
https://github.com/cupy/cupy/blob/e0fe38a8d6a68e53e76c9d99969154d583b89599/tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py#L156

**Comment by leofang** [comment]
If you're comfortable with going ahead and adding tests "in the dark" (meaning with the SciPy PR pending), the decorator that you will add `@testing.with_requires('scipy>=1.x.x')` can help you skip the tests until the supporting SciPy is released. We can then decide whether we wanna merge this PR and then revisit if there's any issue in the tests, or leave it open until the upstream work is done and available for testing.

**Comment by Smit-create** [comment]
All the functions of `cupyx.scipy.ndimage` are now added with `uarray` support. We can add tests and make it look ready in some time.

**Comment by Smit-create** [comment]
The current `cupy` backend on this PR works with the [scipy's ndimage PR](https://github.com/scipy/scipy/pull/14356). A small demo of the working example is implemented in this [colab notebook](https://colab.research.google.com/drive/1Itu2AqrQB2lEbdMgOZtpBM13oW8yOdKp?usp=sharing). I think the PR is ready for a first review.

**Review by toslunar — CHANGES_REQUESTED** [review]
- The colab example LGTM.
- To make the feature public, `with set_backend(cupyx.scipy.ndimage)` looks better than `with set_backend(cupyx.scipy.ndimage._ndimage)`.  I'd follow the pattern in FFT:
https://github.com/cupy/cupy/blob/b477c0295a4fb374f1873dab42951896f8d4685f/cupyx/scipy/fft/__init__.py#L7-L8
- For implementation details, please check my comment at `fourier_gaussian`.  It's optional to address this in the PR.
- I'll review the test soon.

**Comment by Smit-create** [comment]
> * To make the feature public, `with set_backend(cupyx.scipy.ndimage)` looks better than `with set_backend(cupyx.scipy.ndimage._ndimage)`.  I'd follow the pattern in FFT:

Done, thanks! I also updated the colab example accordingly.

**Comment by Smit-create** [comment]
@toslunar  Can you please review the test and if that looks good, I can add more functions to the tests?

File: cupyx/scipy/ndimage/_ndimage.py
Comment (toslunar) [reply cupyx/scipy/ndimage/_ndimage.py:66]: I suggest changes like ed1d75039012a72f4aff400aab10bbc89d49b43a (on https://github.com/cupy/cupy/compare/master...toslunar:uarray-ndimage).  The `_implemented` dict is shared across submodules.
```
>>> import cupyx.scipy.ndimage
>>> name = lambda f: f.__module__+"."+f.__qualname__
>>> {name(k): name(v) for k, v in cupyx.scipy.ndimage._uarray._implemented.items()}
{'scipy.ndimage._filters.correlate': 'cupyx.scipy.ndimage._filters.correlate', 'scipy.ndimage._fourier.fourier_gaussian': 'cupyx.scipy.ndimage._fourier.fourier_gaussian'}
```

**Review by toslunar — CHANGES_REQUESTED** [review]

File: cupyx/scipy/ndimage/_ndimage.py
Comment (Smit-create) [reply cupyx/scipy/ndimage/_ndimage.py:66]: Yeah! That works fine. Looks good. Thanks! I will make that change soon.

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:257]: `TestFilter.test_filter_backend` would be a better place, (or add a new test class).

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:265]: Does the test fail without `_correct_np_dtype`?

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
Comment (toslunar) [inline tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:255]: Given we don't know the exact SciPy version for now, how about the following?
```python
requires_scipy_ndimage_backend = pytest.mark.skip('scipy.ndimage backend feature has not been released')
# TODO: After the feature is released
# requires_scipy_ndimage_backend = testing.with_requires('scipy>=1.x.x')
...
    @requires_scipy_ndimage_backend
    def test_...
```

Comment (Smit-create) [reply tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:255]: That sounds good! Thanks

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
Comment (Smit-create) [reply tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:265]: I wasn't able to test it locally, since I am using mac M1. I tried to pick up related patterns from `fft` tests.

Comment (Smit-create) [reply tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py]: I verified on the `colab`, and tests were passing without `_correct_np_dtype` (the current diff)

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks for working on this @Smit-create! I left one suggestion on the test cases so far (I think they were not using the backend in the GPU case as-is).

The tests here do pass when testing locally (both before and after the suggested change).

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py
Comment (Smit-create) [reply tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:247]: Thanks! that makes sense to me.

**Review by toslunar — APPROVED** [review]
The test LGTM.  Thanks @grlee77 and @Smit-create.

Tests for ndimage functions that are not in `_filters.py` can be added later.

**Comment by Smit-create** [comment]
The tests fail because of a version mismatch in scipy. `sum_labels` was introduced in v1.6.0 and tests are using scipy v.1.4.1. See https://github.com/scipy/scipy/pull/12242

**Comment by grlee77** [comment]
> The tests fail because of a version mismatch in scipy. sum_labels was introduced in v1.6.0 and tests are using scipy v.1.4.1. See https://github.com/scipy/scipy/pull/12242

You can add the following decorator to only test with 1.6 or above:
```Python
@testing.with_requires('scipy>=1.6')
```
(there are examples of this in `test_interpolation.py` and elsewhere)

**Comment by Smit-create** [comment]
> You can add the following decorator to only test with 1.6 or above:

We are already skipping the tests using the decorator `@requires_scipy_ndimage_backend`. I guess the failure arises while building `cupy` from `_uarray.py` when it can't find `sum_labels`.

**Comment by jakirkham** [comment]
cc @GenevieveBuckley (for awareness)

**Comment by kmaehashi** [comment]
Closing as discussed above, let's resume once the implementation lands in SciPy. :+1:

**Pull request closed without merging**
